File: src/okbay/locate.py
from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path
from typing import Any

from .wiki import get_page
from . import paths


def _which(name: str) -> str | None:
    return shutil.which(name)


def _spawn(cmd: list[str]) -> bool:
    """Start ``cmd`` detached; return True if spawn succeeded."""
    try:
        subprocess.Popen(
            cmd,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        return True
    except OSError:
        return False


def _reveal_candidates(target: Path) -> list[list[str]]:
    """Ordered file-manager launch commands for Omarchy Nautilus → xdg-open.

    Omarchy ships Nautilus as the default file browser (``org.gnome.Nautilus``)
    and wraps launches with ``uwsm-app``. Prefer selecting an existing file
    (``nautilus --select``) so Reveal highlights it; otherwise open the parent
    directory. Fall back to ``xdg-open`` so non-Omarchy / headless hosts keep
    working when Nautilus is absent.
    """
    path = target
    parent = path.parent if path.parent.exists() else None
    is_file = path.is_file()
    is_dir = path.is_dir()
    open_dir = path if is_dir else parent
    if open_dir is None or not open_dir.exists():
        return []

    nautilus = _which("nautilus")
    uwsm = _which("uwsm-app")
    xdg = _which("xdg-open")
    cmds: list[list[str]] = []

    def add(cmd: list[str]) -> None:
        cmds.append(cmd)

    if nautilus:
        if is_file:
            select = [nautilus, "--select", str(path)]
            if uwsm:
                add([uwsm, "--", *select])
            add(select)
        # Directory (or missing file → parent): new window on that folder.
        win = [nautilus, "--new-window", str(open_dir)]
        if uwsm:
            add([uwsm, "--", *win])
        add(win)

    if xdg:
        add([xdg, str(open_dir)])

    return cmds


def reveal_path(target: str | Path) -> tuple[list[str] | None, str | None]:
    """Try native file browser then xdg-open. Returns (cmd, error)."""
    path = Path(target)
    candidates = _reveal_candidates(path)
    if not candidates:
        return None, "no file manager / path missing"
    last_err = "file manager launch failed"
    for cmd in candidates:
        if _spawn(cmd):
            return cmd, None
        last_err = f"failed to spawn {' '.join(cmd)}"
    return None, last_err


def locate(stem: str, *, reveal: bool | None = None) -> dict[str, Any]:
    """Resolve wiki + vault paths for a page stem.

    ``reveal`` defaults to env ``OKBAY_REVEAL`` (truthy unless ``0``). Pass
    ``reveal=False`` for resolve-only (atlas open / toast); ``True`` to open a
    file manager when available (Omarchy Nautilus preferred, else xdg-open).
    """
    page = get_page(stem)
    if not page:
        return {"ok": False, "error": f"no page {stem}"}
    sources = page.front.get("extracted_from") or page.front.get("sources") or page.sources or []
    if isinstance(sources, str):
        sources = [sources]
    vault = paths.vault()
    resolved: list[str] = []
    existing: list[str] = []
    for src in sources:
        p = Path(str(src))
        if not p.is_absolute():
            # Prefer vault/ (Biocure), then workspace-relative.
            cand = vault / src
            if not cand.exists():
                cand = paths.workspace() / src
            p = cand
        resolved.append(str(p))
        if p.exists():
            existing.append(str(p))
    files = existing or list(getattr(page, "files", []) or [])
    result: dict[str, Any] = {
        "ok": True,
        "stem": page.stem,
        "title": page.title,
        "wiki": str(page.path),
        "sources": resolved,
        "files": existing,
        "kind": page.kind,
    }
    # Prefer an on-disk vault/source file for highlight; else wiki page path.
    target = (existing[0] if existing else None) or (resolved[0] if resolved else str(page.path))
    result["target"] = target
    if _which("hyprctl"):
        result["hypr"] = "available"
    do_reveal = reveal
    if do_reveal is None:
        do_reveal = os.environ.get("OKBAY_REVEAL", "1") != "0"
    if do_reveal:
        cmd, err = reveal_path(target)
        if cmd:
            result["revealed"] = cmd
            # Friendly label for Atlas status toast (cmd may be absolute paths).
            joined = " ".join(cmd)
            if "nautilus" in joined:
                result["reveal_via"] = "nautilus"
            elif "xdg-open" in joined:
                result["reveal_via"] = "xdg-open"
            else:
                result["reveal_via"] = Path(cmd[0]).name
        else:
            result["reveal_error"] = err or "no file manager / path missing"
    else:
        result["revealed"] = None
    return result

File: crates/okbayd/src/locate.rs
use crate::paths;
use crate::wiki::get_page;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};
use std::process::Command;

fn which(bin: &str) -> bool {
    std::env::var_os("PATH")
        .map(|paths| {
            std::env::split_paths(&paths).any(|dir| {
                let p = dir.join(bin);
                p.is_file()
            })
        })
        .unwrap_or(false)
}

fn spawn(cmd: &[String]) -> bool {
    if cmd.is_empty() {
        return false;
    }
    Command::new(&cmd[0])
        .args(&cmd[1..])
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn()
        .is_ok()
}

/// Ordered Omarchy Nautilus → xdg-open launch candidates for Reveal.
fn reveal_candidates(target: &Path) -> Vec<Vec<String>> {
    let parent = target.parent().filter(|p| p.exists());
    let is_file = target.is_file();
    let is_dir = target.is_dir();
    let open_dir: Option<&Path> = if is_dir {
        Some(target)
    } else {
        parent
    };
    let Some(open_dir) = open_dir else {
        return vec![];
    };

    let has_nautilus = which("nautilus");
    let has_uwsm = which("uwsm-app");
    let has_xdg = which("xdg-open");
    let mut cmds: Vec<Vec<String>> = Vec::new();

    let push = |cmds: &mut Vec<Vec<String>>, parts: &[&str]| {
        cmds.push(parts.iter().map(|s| (*s).to_string()).collect());
    };

    if has_nautilus {
        if is_file {
            let path = target.to_string_lossy();
            if has_uwsm {
                push(
                    &mut cmds,
                    &["uwsm-app", "--", "nautilus", "--select", path.as_ref()],
                );
            }
            push(&mut cmds, &["nautilus", "--select", path.as_ref()]);
        }
        let dir = open_dir.to_string_lossy();
        if has_uwsm {
            push(
                &mut cmds,
                &[
                    "uwsm-app",
                    "--",
                    "nautilus",
                    "--new-window",
                    dir.as_ref(),
                ],
            );
        }
        push(&mut cmds, &["nautilus", "--new-window", dir.as_ref()]);
    }

    if has_xdg {
        let dir = open_dir.to_string_lossy();
        push(&mut cmds, &["xdg-open", dir.as_ref()]);
    }

    cmds
}

fn reveal_path(target: &Path) -> (Option<Vec<String>>, Option<String>) {
    let candidates = reveal_candidates(target);
    if candidates.is_empty() {
        return (None, Some("no file manager / path missing".into()));
    }
    let mut last_err = "file manager launch failed".to_string();
    for cmd in candidates {
        if spawn(&cmd) {
            return (Some(cmd), None);
        }
        last_err = format!("failed to spawn {}", cmd.join(" "));
    }
    (None, Some(last_err))
}

pub fn locate(ws: &Path, stem: &str, reveal: bool) -> Value {
    let page = match get_page(&paths::wiki(ws), stem) {
        Some(p) => p,
        None => return json!({"ok": false, "error": format!("no page {stem}")}),
    };
    let vault = paths::vault(ws);
    let mut resolved = vec![];
    let mut existing = vec![];
    for src in &page.sources {
        let p = Path::new(src);
        let full = if p.is_absolute() {
            p.to_path_buf()
        } else {
            let cand = vault.join(src);
            if cand.exists() {
                cand
            } else {
                ws.join(src)
            }
        };
        let s = full.to_string_lossy().to_string();
        if full.exists() {
            existing.push(s.clone());
        }
        resolved.push(s);
    }
    let target = existing
        .first()
        .cloned()
        .or_else(|| resolved.first().cloned())
        .unwrap_or_else(|| page.path.to_string_lossy().to_string());
    let mut out = json!({
        "ok": true,
        "stem": page.stem,
        "title": page.title,
        "wiki": page.path.to_string_lossy(),
        "sources": resolved,
        "files": existing,
        "kind": page.kind,
        "target": target,
        "revealed": Value::Null,
    });
    if which("hyprctl") {
        out["hypr"] = json!("available");
    }
    if reveal {
        let (cmd, err) = reveal_path(PathBuf::from(&target).as_path());
        if let Some(cmd) = cmd {
            let joined = cmd.join(" ");
            let via = if joined.contains("nautilus") {
                "nautilus"
            } else if joined.contains("xdg-open") {
                "xdg-open"
            } else {
                Path::new(cmd.first().map(|s| s.as_str()).unwrap_or("unknown"))
                    .file_name()
                    .and_then(|s| s.to_str())
                    .unwrap_or("unknown")
            };
            out["revealed"] = json!(cmd);
            out["reveal_via"] = json!(via);
        } else if let Some(e) = err {
            out["reveal_error"] = json!(e);
        } else {
            out["reveal_error"] = json!("no file manager / path missing");
        }
    }
    out
}
